bigquery/tests/unit/test_client.py

```diff
@@ -5161,3 +5161,84 @@ def test__do_multipart_upload_wrong_size(self):
 
         with pytest.raises(ValueError):
             client._do_multipart_upload(file_obj, {}, file_obj_len + 1, None)
+
+    def test__schema_from_json(self):
+        from google.cloud.bigquery.schema import SchemaField
+
+        file_content = """[
+          {
+            "description": "quarter",
+            "mode": "REQUIRED",
+            "name": "qtr",
+            "type": "STRING"
+          },
+          {
+            "description": "sales representative",
+            "mode": "NULLABLE",
+            "name": "rep",
+            "type": "STRING"
+          },
+          {
+            "description": "total sales",
+            "mode": "NULLABLE",
+            "name": "sales",
+            "type": "FLOAT"
+          }
+        ]"""
+        expected = list()
+        json_data = json.loads(file_content)
+
+        for field in json_data:
+            schema_field = SchemaField.from_api_repr(field)
+            expected.append(schema_field)
+
+        client = self._make_client()
+        mock_file_path = "/mocked/file.json"
+
+        open_patch = mock.patch(
+            "builtins.open", new=mock.mock_open(read_data=file_content)
+        )
+        with open_patch as _mock_file:
+            actual = client.schema_from_json(mock_file_path)
+            _mock_file.assert_called_once_with(mock_file_path)
```

Comment:
Check that `close` is called once as well. Ditto for `schema_to_json` test with file path.

Reply:
I made changes to the tests to reflect closing as requested. Used __exit__ for the context manager.

Reply:
Use `json.dump` and a `BytesIO` project will work.

👌 regarding `__exit__`.

Reply:
I tried to use a `BytesIO` object and kept getting this error when using `json.dump`:

```
>           fp.write(chunk)
E           TypeError: a bytes-like object is required, not 'str'

/usr/local/opt/pyenv/versions/3.7.2/lib/python3.7/json/__init__.py:180: TypeError
```

I checked the [docs](https://docs.python.org/3/library/json.html#basic-usage) for `json.dump` and found this:

> The json module always produces str objects, not bytes objects. Therefore, fp.write() must support str input.

I then checked the [docs](https://docs.python.org/3/library/io.html#io.BufferedIOBase.write) for `io.BytesIO` but it doesn't say anything about supporting string input.

> Write the given bytes-like object, b, and return the number of bytes written (always equal to the length of b in bytes, since if the write fails an OSError will be raised).

Do you have an example of what I should be doing?